src/sea.ts

```diff
@@ -0,0 +1,253 @@
+
+  // signtool.js
+  const binPath = bin || process.execPath;
+  const script = SEA_MAIN_SCRIPT
+    .replace('%PATH_TO_BIN%', escapeMaybe(binPath))
+    .replace('%PATH_TO_SCRIPT%', escapeMaybe(receiverPath))
+    .replace('%WINDOWS_SIGN_OPTIONS%', JSON.stringify(options.windowsSign));
+
+  await fs.outputFile(path.join(dir, FILENAMES.SEA_MAIN), script);
+}
+
+async function createBlob(options: InternalSeaOptions) {
+  const args = ['--experimental-sea-config', 'sea-config.json'];
+  const bin = process.execPath;
+  const cwd = options.dir;
+
+  log(`Calling ${bin} with options:`, args);
+
+  return spawnPromise(bin, args, {
+    cwd
+  });
+}
+
+async function createBinary(options: InternalSeaOptions) {
+  const { dir, filename } = options;
+
+  log(`Creating ${filename} in ${dir}`);
+
+  // Copy Node over
+  const seaPath = path.join(dir, filename);
+  await fs.copyFile(process.execPath, seaPath);
+
+  // Remove the Node signature
+  const signtool = path.join(__dirname, '../../vendor/signtool.exe');
+  await spawnPromise(signtool, [
+    'remove',
+    '/s',
+    seaPath
+  ]);
+
+  // Inject the blob
+  const blob = await fs.readFile(path.join(dir, FILENAMES.SEA_BLOB));
+  await postject.inject(seaPath, 'NODE_SEA_BLOB', blob, {
+    sentinelFuse: 'NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2'
+  });
+}
+
+async function cleanup(options: InternalSeaOptions) {
+  const { dir } = options;
+  const toRemove = [
+    FILENAMES.SEA_BLOB,
+    FILENAMES.SEA_MAIN,
+    FILENAMES.SEA_CONFIG
+  ];
+
+  for (const file of toRemove) {
+    try {
+      await fs.remove(path.join(dir, file));
+    } catch (error) {
+      console.warn(`Tried and failed to remove ${file}. Continuing.`, error);
+    }
+  }
+}
+
+async function getOptions(options: Partial<SeaOptions>): Promise<InternalSeaOptions> {
+  const cloned = { ...options };
+
+  if (!cloned.path) {
+    cloned.path = path.join(os.homedir(), '.electron', 'windows-sign', 'sea.exe');
+    await fs.mkdirp(cloned.path);
+  }
+
+  if (!cloned.bin) {
+    cloned.bin = process.execPath;
+  }
+
+  if (!cloned.windowsSign) {
+    throw new Error('Did not find windowsSign options, which are required');
+  }
+
+  return {
+    path: cloned.path,
+    dir: path.dirname(cloned.path),
+    filename: path.basename(cloned.path),
+    bin: cloned.bin,
+    windowsSign: cloned.windowsSign
+  };
+}
+
+/**
+ * Ensures that the current Node.js version supports SEA app generation and errors if not.
+ */
+function checkCompatibility() {
+  const version = process.versions.node;
+  const split = version.split('.');
+  const major = parseInt(split[0], 10);
+  const minor = parseInt(split[1], 10);
+
+  if (major >= 20) {
+    return true;
+  }
+
+  if (major === 19 && minor >= 7) {
+    return true;
+  }
+
+  if (major === 18 && minor >= 16) {
+    return true;
+  }
+
+  throw new Error(`Your Node.js version (${process.version}) does not support Single Executable Applications. Please upgrade your version of Node.js.`);

```

Comment:
Since you mentioned this will be consumed by `electron-winstaller`: that module is currently on Node 8, so we might need to update it (and potentially Forge, which is on Node 16 and has `electron-winstaller` as an optional dep - not sure how we handle these) to 18 first.

Reply:
It's a little tough to do with backwards compat - since only a small number of people will actually need to use the `sea` ability (apps with EV certificates and custom codesigning pipelines in the cloud), I think my recommendation would be:

- Keep `@electron/windows-installer` at Node 8. Or upgrade it, but maybe only to ~16 or so.
- Document that `windowsSign` options inside `@electron/windows-installer` requires at least Node 18.